File: lib/services/trakt/trakt_shows_api.dart
import 'package:dio/dio.dart';
import '../../core/config/trakt_config.dart';
import '../../models/trakt/trakt_show.dart';

class TraktShowsApi {
  final Dio _dio;
  
  TraktShowsApi(this._dio);
  
  Future<List<TraktShow>> getTrending({
    int page = 1,
    int limit = 10,
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/shows/trending',
      queryParameters: {
        'page': page,
        'limit': limit,
        if (extended != null) 'extended': extended.value,
      },
    );
    
    return (response.data as List)
        .map((json) => TraktShow.fromJson(json['show']))
        .toList();
  }
  
  Future<List<TraktShow>> getPopular({
    int page = 1,
    int limit = 10,
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/shows/popular',
      queryParameters: {
        'page': page,
        'limit': limit,
        if (extended != null) 'extended': extended.value,
      },
    );
    
    return (response.data as List)
        .map((json) => TraktShow.fromJson(json))
        .toList();
  }
  
  Future<TraktShow> getSummary(
    String traktSlug, {
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/shows/$traktSlug',
      queryParameters: {
        if (extended != null) 'extended': extended.value,
      },
    );
    
    return TraktShow.fromJson(response.data);
  }
  
  Future<List<TraktShow>> getRelated(
    String showId, {
    int page = 1,
    int limit = 10,
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/shows/$showId/related',
      queryParameters: {
        'page': page,
        'limit': limit,
        if (extended != null) 'extended': extended.value,
      },
    );
    
    return (response.data as List)
        .map((json) => TraktShow.fromJson(json))
        .toList();
  }
  
  Future<List<Map<String, dynamic>>> getMyAiringShows({
    String? startDate,
    int days = 7,
  }) async {
    final start = startDate ?? DateTime.now().toIso8601String().split('T')[0];
    
    final response = await _dio.get(
      '/calendars/my/shows/$start/$days',
      queryParameters: {
        'extended': 'full',
      },
    );
    
    return (response.data as List)
        .map((json) => json as Map<String, dynamic>)
        .toList();
  }
  
  Future<List<Map<String, dynamic>>> getMyNewShows({
    String? startDate,
    int days = 7,
  }) async {
    final start = startDate ?? DateTime.now().toIso8601String().split('T')[0];
    
    final response = await _dio.get(
      '/calendars/my/shows/new/$start/$days',
      queryParameters: {
        'extended': 'full',
      },
    );
    
    return (response.data as List)
        .map((json) => json as Map<String, dynamic>)
        .toList();
  }
  
  Future<List<Map<String, dynamic>>> getMySeasonPremieres({
    String? startDate,
    int days = 7,
  }) async {
    final start = startDate ?? DateTime.now().toIso8601String().split('T')[0];
    
    final response = await _dio.get(
      '/calendars/my/shows/premieres/$start/$days',
      queryParameters: {
        'extended': 'full',
      },
    );
    
    return (response.data as List)
        .map((json) => json as Map<String, dynamic>)
        .toList();
  }
  
  Future<List<Map<String, dynamic>>> getMyFinales({
    String? startDate,
    int days = 7,
  }) async {
    final start = startDate ?? DateTime.now().toIso8601String().split('T')[0];
    
    final response = await _dio.get(
      '/calendars/my/shows/finales/$start/$days',
      queryParameters: {
        'extended': 'full',
      },
    );
    
    return (response.data as List)
        .map((json) => json as Map<String, dynamic>)
        .toList();
  }
  
  Future<List<Map<String, dynamic>>> getAllNewShows({
    String? startDate,
    int days = 7,
  }) async {
    final start = startDate ?? DateTime.now().toIso8601String().split('T')[0];
    
    final response = await _dio.get(
      '/calendars/all/shows/new/$start/$days',
      queryParameters: {
        'extended': 'full',
      },
    );
    
    return (response.data as List)
        .map((json) => json as Map<String, dynamic>)
        .toList();
  }
  
  Future<List<TraktShow>> getAnticipated({
    int page = 1,
    int limit = 10,
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/shows/anticipated',
      queryParameters: {
        'page': page,
        'limit': limit,
        if (extended != null) 'extended': extended.value,
      },
    );
    
    return (response.data as List)
        .map((json) => TraktShow.fromJson(json['show']))
        .toList();
  }
  
  Future<List<TraktShow>> getWatched({
    String period = 'weekly',
    int page = 1,
    int limit = 10,
    TraktExtended? extended,
  }) async {
    final response = await _dio.get(
      '/shows/watched/$period',
      queryParameters: {
        'page': page,
        'limit': limit,
        if (extended != null) 'extended': extended.value,
      },
    );
    
    return (response.data as List)
        .map((json) => TraktShow.fromJson(json['show']))
        .toList();
  }
}

File: lib/trakt/api/TraktShowsApi.kt
package app.moviebase.trakt.api

import app.moviebase.trakt.TraktExtended
import app.moviebase.trakt.core.endPoint
import app.moviebase.trakt.core.parameterExtended
import app.moviebase.trakt.core.parameterLimit
import app.moviebase.trakt.core.parameterPage
import app.moviebase.trakt.model.TraktAnticipatedShow
import app.moviebase.trakt.model.TraktRating
import app.moviebase.trakt.model.TraktShow
import app.moviebase.trakt.model.TraktTrendingShow
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.HttpRequestBuilder
import io.ktor.client.request.get
import io.ktor.client.request.parameter

class TraktShowsApi(
    private val client: HttpClient,
) {
    suspend fun getTrending(
        page: Int,
        limit: Int,
        extended: TraktExtended? = null,
    ): List<TraktTrendingShow> =
        client
            .get {
                endPointShows("trending")
                parameterPage(page)
                parameterLimit(limit)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getPopular(
        page: Int,
        limit: Int,
        extended: TraktExtended? = null,
    ): List<TraktShow> =
        client
            .get {
                endPointShows("popular")
                parameterPage(page)
                parameterLimit(limit)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getAnticipated(
        page: Int,
        limit: Int,
        extended: TraktExtended? = null,
    ): List<TraktAnticipatedShow> =
        client
            .get {
                endPointShows("anticipated")
                parameterPage(page)
                parameterLimit(limit)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getPlayed(
        page: Int,
        limit: Int,
        extended: TraktExtended? = null,
    ): List<TraktShow> =
        client
            .get {
                endPointShows("played")
                parameterPage(page)
                parameterLimit(limit)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getWatched(
        page: Int,
        limit: Int,
        extended: TraktExtended? = null,
    ): List<TraktShow> =
        client
            .get {
                endPointShows("watched")
                parameterPage(page)
                parameterLimit(limit)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getCollected(
        page: Int,
        limit: Int,
        extended: TraktExtended? = null,
    ): List<TraktShow> =
        client
            .get {
                endPointShows("collected")
                parameterPage(page)
                parameterLimit(limit)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getRelated(
        showId: String,
        page: Int,
        limit: Int,
        extended: TraktExtended? = null,
    ): List<TraktShow> =
        client
            .get {
                endPointShow(showId, "related")
                parameterPage(page)
                parameterLimit(limit)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getSummary(
        showId: String,
        extended: TraktExtended? = null,
    ): TraktShow =
        client
            .get {
                endPointShows(showId)
                extended?.let { parameterExtended(it) }
            }.body()

    suspend fun getRating(traktSlug: String): TraktRating =
        client
            .get {
                endPointShow(traktSlug, "ratings")
            }.body()

    suspend fun getStats(traktSlug: String): TraktRating =
        client
            .get {
                endPointShow(traktSlug, "stats")
            }.body()

    suspend fun getTranslations(
        showId: String,
        language: String? = null,
    ): List<TraktShow> =
        client
            .get {
                endPointShow(showId, "translations", language ?: "")
            }.body()

    suspend fun getComments(
        showId: String,
        sort: String = "newest",
        page: Int = 1,
        limit: Int = 10,
    ): List<TraktShow> =
        client
            .get {
                endPointShow(showId, "comments", sort)
                parameterPage(page)
                parameterLimit(limit)
            }.body()

    suspend fun getProgress(
        showId: String,
        hidden: Boolean = false,
        specials: Boolean = false,
        countSpecials: Boolean = true,
    ): TraktShow =
        client
            .get {
                endPointShow(showId, "progress", "watched")
                parameter("hidden", hidden)
                parameter("specials", specials)
                parameter("count_specials", countSpecials)
            }.body()

    private fun HttpRequestBuilder.endPointShow(
        traktSlug: String,
        vararg paths: String,
    ) {
        endPoint("shows", traktSlug, *paths)
    }

    private fun HttpRequestBuilder.endPointShows(vararg paths: String) {
        endPoint("shows", *paths)
    }
}
